third_party/proton/csrc/lib/Profiler/CuptiProfiler.cpp

```diff
@@ -199,17 +237,66 @@ void CuptiProfiler::CuptiProfilerPimpl::callbackFn(void *userData,
-  if (callbackData->callbackSite == CUPTI_API_ENTER) {
-    auto scopeId = Scope::getNewScopeId();
-    profilerState.record(scopeId);
-    profilerState.enterOp(scopeId);
-    profiler.correlation.correlate(callbackData->correlationId);
-  } else if (callbackData->callbackSite == CUPTI_API_EXIT) {
-    profilerState.exitOp();
-    profiler.correlation.submit(callbackData->correlationId);
+  CuptiProfiler &profiler = threadState.profiler;
+  if (domain == CUPTI_CB_DOMAIN_RESOURCE) {
+    auto *resourceData =
+        reinterpret_cast<CUpti_ResourceData *>(const_cast<void *>(cbData));
+    auto *graphData =
+        reinterpret_cast<CUpti_GraphData *>(resourceData->resourceDescriptor);
+    auto *pImpl = dynamic_cast<CuptiProfilerPimpl *>(profiler.pImpl.get());
+    uint32_t graphId = 0;
+    uint32_t graphExecId = 0;
+    if (graphData->graph)
+      cupti::getGraphId<true>(graphData->graph, &graphId);
+    if (graphData->graphExec)
+      cupti::getGraphExecId<true>(graphData->graphExec, &graphExecId);
+    if (cbId == CUPTI_CBID_RESOURCE_GRAPHNODE_CREATED ||
+        cbId == CUPTI_CBID_RESOURCE_GRAPHNODE_CLONED) {
+      if (!pImpl->graphIdToNumInstances.contain(graphId))
+        pImpl->graphIdToNumInstances[graphId] = 1;
+      else
+        pImpl->graphIdToNumInstances[graphId]++;
+    } else if (cbId == CUPTI_CBID_RESOURCE_GRAPHNODE_DESTROY_STARTING) {
+      pImpl->graphIdToNumInstances[graphId]--;
+    } else if (cbId == CUPTI_CBID_RESOURCE_GRAPHEXEC_CREATED) {
+      pImpl->graphExecIdToGraphId[graphExecId] = graphId;
+    } else if (cbId == CUPTI_CBID_RESOURCE_GRAPHEXEC_DESTROY_STARTING) {
+      pImpl->graphExecIdToGraphId.erase(graphExecId);
+    } else if (cbId == CUPTI_CBID_RESOURCE_GRAPH_DESTROY_STARTING) {
+      pImpl->graphIdToNumInstances.erase(graphId);
+    }
+  } else {
+    const CUpti_CallbackData *callbackData =
+        reinterpret_cast<const CUpti_CallbackData *>(cbData);
+    if (callbackData->callbackSite == CUPTI_API_ENTER) {
+      auto scopeId = Scope::getNewScopeId();
+      threadState.record(scopeId);
+      threadState.enterOp(scopeId);
+      size_t numInstances = 1;
+      if (cbId == CUPTI_DRIVER_TRACE_CBID_cuGraphLaunch ||
+          cbId == CUPTI_DRIVER_TRACE_CBID_cuGraphLaunch_ptsz) {
+        auto *pImpl = dynamic_cast<CuptiProfilerPimpl *>(profiler.pImpl.get());
+        auto graphExec = reinterpret_cast<const cuGraphLaunch_params *>(
+                             callbackData->functionParams)
+                             ->hGraph;
+        uint32_t graphExecId = 0;
+        cupti::getGraphExecId<true>(graphExec, &graphExecId);
+        // It's possible that the graph is not captured, because proton starts
+        // profiling after the graph is created. In this case, we cannot get the
+        // number of instances belonging to the graph and thus cannot delete the
+        // map entry.
+        numInstances = std::numeric_limits<size_t>::max();
```

Comment:
Is it worth issuing a warning to the user about this? This seems like atypical usage and if it's happening may actually not be what the user wants and just don't know it's happening (if the profiling is launched with a script and/or within some other framework).

Reply:
Sure. Good idea. I add a `std::cerr` message. Let me know if you have other thoughts.